### PR TITLE
Fixed incorrect prompt formatting for chat based ReAct agent

### LINKED ISSUE #5390: Incorrect prompt formatting when initializing ReActDocstoreAgent with a chat model
### System Info

Langchain v0.0.183 

### Who can help?

@hwchase17 
@agola11

### Information

- [X] The official example notebooks/scripts
- [ ] My own modified scripts

### Related Components

- [X] LLMs/Chat Models
- [ ] Embedding Models
- [X] Prompts / Prompt Templates / Prompt Selectors
- [ ] Output Parsers
- [ ] Document Loaders
- [ ] Vector Stores / Retrievers
- [ ] Memory
- [ ] Agents / Agent Executors
- [ ] Tools / Toolkits
- [ ] Chains
- [ ] Callbacks/Tracing
- [ ] Async

### Reproduction

```python
from langchain import Wikipedia
from langchain.chat_models import ChatOpenAI
from langchain.agents import initialize_agent, Tool
from langchain.agents import AgentType
from langchain.agents.react.base import DocstoreExplorer
docstore=DocstoreExplorer(Wikipedia())
tools = [
    Tool(
        name="Search",
        func=docstore.search,
        description="useful for when you need to ask with search"
    ),
    Tool(
        name="Lookup",
        func=docstore.lookup,
        description="useful for when you need to ask with lookup"
    )
]


llm = ChatOpenAI()
react = initialize_agent(tools, llm, agent=AgentType.REACT_DOCSTORE, verbose=True)

question = 'Question'
react.run(question)
```
Running this snippet initializes a ReActDocStoreAgent and runs it on the given question.

[initialize_agent( )](https://github.com/hwchase17/langchain/blob/master/langchain/agents/initialize.py#L12) returns an AgentExecutor. During this call an agent is created using the [from_llm_and_tools( )](https://github.com/hwchase17/langchain/blob/master/langchain/agents/agent.py#L528) function. This function creates a new llm_chain for the agent and sets the prompt using agent_cls.create_prompt( ).  Since we are using a ReActDocStoreAgent, [the function](https://github.com/hwchase17/langchain/blob/master/langchain/agents/react/base.py#L35) simply returns the WIKI_PROMPT, which is a PromptTemplate object.

[run( )](https://github.com/hwchase17/langchain/blob/master/langchain/agents/agent.py#L934) from AgentExecutor calls [_take_next_step( )](https://github.com/hwchase17/langchain/blob/master/langchain/agents/agent.py#L748) which calls agent.plan( ).

[plan( )](https://github.com/hwchase17/langchain/blob/master/langchain/agents/agent.py#L425) eventually calls llm_chain.predict( ). Eventually llm_chain calls [generate( )](https://github.com/hwchase17/langchain/blob/master/langchain/chains/llm.py#L72). This function creates a prompt by calling [prep_prompt( )](https://github.com/hwchase17/langchain/blob/master/langchain/chains/llm.py#L94). The prompt is a StringPromptValue since it is created using [format_prompt( )](https://github.com/hwchase17/langchain/blob/master/langchain/prompts/base.py#L230).
This variable is then used as an argument for the function llm.generate_prompt( ). [generate_prompt( )](https://github.com/hwchase17/langchain/blob/master/langchain/chat_models/base.py#L136) converts the prompt to message format using [to_messages( )](https://github.com/hwchase17/langchain/blob/master/langchain/prompts/base.py#L98), which treats the entire prompt as a HumanMessage. 

### Expected behavior

The issue here is that the prompt includes both human and AI messages. Ideally, the questions and observations should be human messages whereas the thoughts and actions should be AI messages. Treating the entire prompt as a human message may decrease prompt quality and lead to suboptimal performance.

### PR DESCRIPTION
**Pull request by oguzgultepe**
# Fixed incorrect prompt formatting for chat based ReAct agent

Fixes #5390
Created two new classes (ReActStringPromptValue and ReActPromptTemplate) in order to handle correct prompt formatting for chat based models.

My twitter handle: @gultepeoguz

## Before submitting
Format/lint/tests already run.

## Who can review?

Community members can review the PR once tests pass. Tag maintainers/contributors who might be interested:

### CODE REVIEW

**Comment by oguzgultepe** [comment]
@hwchase17 The tests were failing because of a match case statement which is only available python>=3.10.
Replaced it with if statements so now it should be good.

**Review by hwchase17 — COMMENTED** [review]
if we are going to do this properly, better to have a method where we construct the messages directly (rather than construct a string and parse out sections)

**Comment by oguzgultepe** [comment]
@hwchase17 I agree that constructing messages directly would be a better approach. It would be good to store examples as they are and construct string/message when called explicitly. However, when we create a PromptTemplate using [from_examples](https://github.com/hwchase17/langchain/blob/master/langchain/prompts/prompt.py#L79) method, examples are converted to string format (with separators etc.) and stored as such.
I can change the PromptTemplate class to store examples as they are and generate string/messages when explicitly called. I can also add optional 'user_prefixes', 'agent_prefixes' variables to the constructor so that the class is applicable to other prompting methods as well.
I would suggest completing this pull request and opening a new issue in order to keep the scope of the changes minimal and focused. Let me know what you think.

**Comment by oguzgultepe** [comment]
@hwchase17

**Comment by oguzgultepe** [comment]
@agola11

File: langchain/agents/react/prompt.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+from langchain.schema import AIMessage, BaseMessage, HumanMessage
+
+
+class ReActStringPromptValue(StringPromptValue):
+    """PromptValue to be returned from ReActPromptTemplate"""
+
+    def to_messages(self) -> List[BaseMessage]:
+        # Matches the first word
+        prefix_regex = re.compile(r"^.+?\b")
+        thought = ""
```
Comment (pamelafox) [inline langchain/agents/react/prompt.py:16]: this variable is unused, I assume due to tought being typo'd later

**Comment by cutmasta-kun** [comment]
How is the state of this PR?
I think beeing able to use the ChatOpenAI LLM with AgentType.REACT_DOCSTORE would be awesome.
Also the support for OpenAI Models like text-davinci-003 will end in Feb 2024.

File: langchain/agents/react/prompt.py (hunk 1 above)
Comment (oguzgultepe) [reply langchain/agents/react/prompt.py:16]: yes there is a typo on the later uses of the variable

**Comment by oguzgultepe** [comment]
@pamelafox typo resolved
it should be ready to merge now as it is

**Comment by leo-gan** [comment]
@hwchase17 FYI

**Comment by efriis** [comment]
Hey @oguzgultepe ! Sorry for the delay here. I agree with Harrison that it would be better to do this the "right" way rather than introduce new types that handle conversion of a specific string format that we shouldn't have around forever.

If you want to switch the PR to do that here, you're welcome to! Otherwise I will close it tomorrow (and can reopen if you see this later 😊)

**Pull request closed without merging**